MapboxNavigation/NavigationMapView.swift

```diff
@@ -709,6 +720,190 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
     }
 
     // MARK: - Vanishing route line methods
+
+    var primaryRoutePoints: RoutePoints?
+    var primaryRouteLineGranularDistances: RouteLineGranularDistances?
+    var primaryRouteRemainingDistancesIndex: Int?
+    var routeLineTracksTraversal: Bool = false
+    private var startDate = Date()
+    public var fractionTraveled: Double = 0.0
+    var preFractionTraveled: Double = 0.0
+    var timer: Timer? = nil
+    
+    struct RoutePoints {
+        var nestedList: [[[CLLocationCoordinate2D]]]
+        var flatList: [CLLocationCoordinate2D]
+    }
+    
+    struct RouteLineGranularDistances {
+        var distance: Double
+        var distanceArray: [RouteLineDistancesIndex]
+    }
+    
+    struct RouteLineDistancesIndex {
+        var point: CLLocationCoordinate2D
+        var distanceRemaining: Double
+    }
+    
+    func initPrimaryRoutePoints(route: Route) {
+        primaryRoutePoints = parseRoutePoints(route: route)
+        primaryRouteLineGranularDistances = calculateRouteGranularDistances(coordinates: primaryRoutePoints?.flatList ?? [])
+    }
+    
+    /**
+    Tranform the route data into nested arrays of legs -> steps -> coordinates.
+    The first and last point of adjacent steps overlap and are duplicated.
+    */
+    func parseRoutePoints(route: Route) -> RoutePoints {
```

Comment:
Would it be possible to replace current method with one liner like this?
```
let routeCoordinates = route.legs.compactMap({ $0.steps.compactMap({ $0.shape?.coordinates }).reduce([], +) }).reduce([], +)
```

Reply:
We can also implement unit-test for this method to make sure it correctly creates list of coordinates.

Reply:
I tried to use one line but it may have some problem with the nested array. Yes, I've tested it with playground and will add the unit test to this function.